File: DSA/src/Questions/StackQueue/BasicCalculator.java
package Questions.StackQueue;

import java.util.Stack;

/*
 Given a string s representing a valid expression, implement a basic calculator to evaluate it, and return the result of the evaluation.

Note: You are not allowed to use any built-in function which evaluates strings as mathematical expressions, such as eval().

 

Example 1:

Input: s = "1 + 1"
Output: 2
Example 2:

Input: s = " 2-1 + 2 "
Output: 3
Example 3:

Input: s = "(1+(4+5+2)-3)+(6+8)"
Output: 23
 

Constraints:

1 <= s.length <= 3 * 105
s consists of digits, '+', '-', '(', ')', and ' '.
s represents a valid expression.
'+' is not used as a unary operation (i.e., "+1" and "+(2 + 3)" is invalid).
'-' could be used as a unary operation (i.e., "-1" and "-(2 + 3)" is valid).
There will be no two consecutive operators in the input.
Every number and running calculation will fit in a signed 32-bit integer.
Time Complexity: O(n)
Space Complexity: O(1)
 */
public class BasicCalculator {
    public boolean isDigit(char currentChar) {
        return '0' <= currentChar && currentChar <= '9';
    }

    public int calculate(String s) {
        Stack<Integer> stack = new Stack<>();
        int result = 0;
        int sign = 1;
        int i = 0;

        while (i < s.length()) {
            char c = s.charAt(i);

            if (isDigit(c)) {
                int num = 0;
                while (i < s.length() && isDigit(s.charAt(i))) {
                    num = num * 10 + (s.charAt(i) - '0');
                    i++;
                }
                result += sign * num;
                i--; // Step back to process the next character correctly
            } else if (c == '+') {
                sign = 1;
            } else if (c == '-') {
                sign = -1;
            } else if (c == '(') {
                // Push the current result and sign onto the stack
                stack.push(result);
                stack.push(sign);
                // Reset result and sign for new sub-expression
                result = 0;
                sign = 1;
            } else if (c == ')') {
                // Pop the sign and result before the parenthesis
                result = stack.pop() * result + stack.pop();
            }
            i++;
        }

        return result;
    }

    int i = 0;

    public int calculateOP(String s) {
        return calc(s);
    }

    private int calc(String s) {
        int num = 0, ans = 0;
        int sign = 1;
        while (i < s.length()) {
            char ch = s.charAt(i++);
            if (ch >= '0' && ch <= '9') {
                num = num * 10 + ch - '0';
            } else if (ch == '(') {
                num = calc(s);
            } else if (ch == ')') {
                return ans += num * sign;
            } else if (ch == '+' || ch == '-') {
                ans += num * sign;
                num = 0;
                sign = (ch == '-') ? -1 : 1;
            }
        }
        ans += num * sign;
        return ans;
    }

    public static void main(String[] args) {
        BasicCalculator basicCalculator = new BasicCalculator();
        System.out.println(basicCalculator.calculate("1 + 1")); // Output: 2
        System.out.println(basicCalculator.calculate(" 2-1 + 2 ")); // Output: 3
        System.out.println(basicCalculator.calculate("(1+(4+5+2)-3)+(6+8)")); // Output: 23
    }
}

File: DSA/Python/Questions/StackQueue/BasicCalculator.py
"""
# Question: Basic Calculator
# Link: https://leetcode.com/problems/basic-calculator/

# Evaluate mathematical expressions with parentheses

# Time Complexity: O(n)
# Space Complexity: O(n)

# Algorithm:
# 1. Process each character in expression
# 2. Handle numbers, operators, and parentheses
# 3. Use stack for nested expressions
# 4. Return final evaluation result

# Key Components:
# - calculate(): Main expression evaluator
# - Stack for tracking nested calculations
# - Sign handling for negative numbers
"""


class BasicCalculator:
    def is_digit(self, char: str) -> bool:
        return "0" <= char <= "9"

    def calculate(self, s: str) -> int:
        stack = []
        result = 0
        sign = 1
        i = 0

        while i < len(s):
            char = s[i]

            if self.is_digit(char):
                num = 0
                while i < len(s) and self.is_digit(s[i]):
                    num = num * 10 + int(s[i])
                    i += 1
                result += sign * num
                i -= 1
            elif char == "+":
                sign = 1
            elif char == "-":
                sign = -1
            elif char == "(":
                stack.append(result)
                stack.append(sign)
                result = 0
                sign = 1
            elif char == ")":
                result = stack.pop() * result + stack.pop()
            i += 1

        return result

    def calculate_recursive(self, s: str) -> int:
        self.i = 0
        return self.calc(s)

    def calc(self, s: str) -> int:
        num = 0
        ans = 0
        sign = 1

        while self.i < len(s):
            char = s[self.i]
            self.i += 1

            if self.is_digit(char):
                num = num * 10 + int(char)
            elif char == "(":
                num = self.calc(s)
            elif char == ")":
                return ans + num * sign
            elif char in "+-":
                ans += num * sign
                num = 0
                sign = 1 if char == "+" else -1

        ans += num * sign
        return ans


def main():
    calculator = BasicCalculator()
    print(calculator.calculate("1 + 1"))  # Output: 2
    print(calculator.calculate(" 2-1 + 2 "))  # Output: 3
    print(calculator.calculate("(1+(4+5+2)-3)+(6+8)"))  # Output: 23


if __name__ == "__main__":
    main()
